### PR TITLE
fix(auth): Fix for checking current preferred type before setting the MFAType as enabled to ensure the current preference is not cleared out

### PR DESCRIPTION
**Pull request by gpanshu**
Fix for checking current preferred type before setting the MFAType as enabled to ensure the current preference is not cleared out

- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:* For updating of mfa preferences we were just updating based on what customer had provided without taking into consideration what current preferences were which meant we could get into scenarios where 2 preferences were set to preferred. Hence in this PR I aim to fix that by prefetching the current preferences before updating it.

*How did you test these changes?*
(Please add a line here how the changes were tested)

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

*General Checklist*
- [x] Added Unit Tests
- [ ] Added Integration Tests
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 1]
```diff
@@ -2223,45 +2224,57 @@ internal class RealAWSCognitoAuthPlugin(
...
-                                            enabled = it.mfaEnabled
-                                            it.mfaPreferred ?.let { preferred -> preferredMfa = preferred }
-                                        }
-                                    }
-                                    this.softwareTokenMfaSettings = totp?.let {
-                                        SoftwareTokenMfaSettingsType.invoke {
-                                            enabled = it.mfaEnabled
-                                            it.mfaPreferred ?.let { preferred -> preferredMfa = preferred }
+        if (sms == null && totp == null) {
+            onError.accept(InvalidParameterException())
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2228]: A few issues here.

If we are going to callback in onError we need to return so we don't continue with another callback.

Also, the default messaging for InvalidParameterException is not very helpful. What does iOS say in this instance?

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 2]
```diff
@@ -2223,45 +2224,57 @@ internal class RealAWSCognitoAuthPlugin(
...
                         }
                     }
-                }
 
-                else -> onError.accept(InvalidStateException())
+                    else -> onError.accept(InvalidStateException())
+                }
             }
-        }
+        }, {})
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2277]: If fetch call fails, you aren't returning success or error back. There should be a test for this case.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 3]
```diff
@@ -2223,45 +2224,57 @@ internal class RealAWSCognitoAuthPlugin(
...
+                                val accessToken = getSession().userPoolTokensResult.value?.accessToken
+                                accessToken?.let { token ->
+                                    authEnvironment
+                                        .cognitoAuthService
+                                        .cognitoIdentityProviderClient
+                                        ?.setUserMfaPreference {
+                                            this.accessToken = token
+                                            this.smsMfaSettings = sms?.let { it ->
+                                                val preferredMFASetting = it.mfaPreferred
+                                                    ?: (userPreference.preferred == MFAType.SMS)
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2245]: Faulty logic. If Disabled is set, we don't want it to remain preferred. Test should catch this.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 4]
```diff
@@ -2223,45 +2224,57 @@ internal class RealAWSCognitoAuthPlugin(
...
+                                                SmsMfaSettingsType.invoke {
+                                                    enabled = it.mfaEnabled
+                                                    preferredMfa = preferredMFASetting
+                                                }
+                                            }
+                                            this.softwareTokenMfaSettings = totp?.let { it ->
+                                                val preferredMFASetting = it.mfaPreferred
+                                                    ?: (userPreference.preferred == MFAType.TOTP)
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2253]: Faulty logic. If Disabled is set, we don't want it to remain preferred. Test should catch this.

**Review by tylerjroach — CHANGES_REQUESTED** [review]

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 3 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt]: done

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 4 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt]: done.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 2 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt]: done

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 1 above)
Comment (tylerjroach) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt]: This is a good change. Lets take the extra step to add in, but improve the default message/recovery suggestion to help the customer here.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 5]
```diff
@@ -2223,45 +2223,56 @@ internal class RealAWSCognitoAuthPlugin(
...
                     }
-                }
 
-                else -> onError.accept(InvalidStateException())
+                    else -> onError.accept(InvalidStateException())
+                }
             }
-        }
+        }, {
+            onError.accept(it)
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2274]: We should be careful with this message being thrown. The message of failure here is if fetch fails, which will be confusing to the customer in an update call. Can we wrap the inner exception in an outer exception that has a message like "Failed to fetch current MFA preference, which is a prerequisite of updating MFA preferences.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 6]
```diff
@@ -2223,45 +2223,56 @@ internal class RealAWSCognitoAuthPlugin(
...
+                                val accessToken = getSession().userPoolTokensResult.value?.accessToken
+                                accessToken?.let { token ->
+                                    authEnvironment
+                                        .cognitoAuthService
+                                        .cognitoIdentityProviderClient
+                                        ?.setUserMfaPreference {
+                                            this.accessToken = token
+                                            this.smsMfaSettings = sms?.let { it ->
+                                                val preferredMFASetting = it.mfaPreferred
+                                                    ?: (userPreference.preferred == MFAType.SMS && it.mfaEnabled)
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2241]: Change looks good. I am curious now though. Without the change to check for mfaEnabled. Will Cognito accept "enabled = false, preferred = true"?

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 5 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt]: done

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 1 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt]: done.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 6 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2246]: this is not even allowed from an enum perspective but if sent we will get an invalidparameterexception

**Review by tylerjroach — APPROVED** [review]

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 7]
```diff
@@ -2223,45 +2224,67 @@ internal class RealAWSCognitoAuthPlugin(
...
+                                            this.softwareTokenMfaSettings = totp?.let { it ->
+                                                val preferredMFASetting = it.mfaPreferred
+                                                    ?: (userPreference.preferred == MFAType.TOTP && it.mfaEnabled)
+                                                SoftwareTokenMfaSettingsType.invoke {
+                                                    enabled = it.mfaEnabled
+                                                    preferredMfa = preferredMFASetting
+                                                }
+                                            }
+                                        }?.also {
+                                            println("Ankit's test $it")
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2261]: This line needs removed

Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt]: done

**Review by tylerjroach — APPROVED** [review]

**Review by mattcreaser — APPROVED** [review]

**Pull request merged**